src/quote/Quote.ts

```diff
@@ -0,0 +1,125 @@
+ */
+
+import Ajv from 'ajv'
+import QuoteSchema from './QuoteSchema'
+import Identity from '../identity/Identity'
+import IQuote, { IQuoteAgreement, IQuoteAttesterSigned } from '../types/Quote'
+import { hashObjectAsStr, verify } from '../crypto/Crypto'
+
+export function validateQuoteSchema(
+  schema: object,
+  validate: object,
+  messages?: string[]
+): boolean | PromiseLike<any> {
+  const ajv = new Ajv()
+  ajv.addMetaSchema(QuoteSchema)
+  const result = ajv.validate(
+    JSON.parse(JSON.stringify(schema)),
+    JSON.parse(JSON.stringify(validate))
+  )
+  if (!result && ajv.errors) {
+    if (messages) {
+      ajv.errors.forEach((error: any) => {
+        messages.push(error.message)
+      })
+    }
+  }
+  return !!result
+}
+
+export function verifyQuoteHash(quote: IQuote, quoteHash: string): boolean {
+  return hashObjectAsStr(quote) === quoteHash
+}
+
+export function fromAttesterSignedInput(
+  deserializedQuote: IQuoteAttesterSigned
+): IQuoteAttesterSigned {
+  const { attesterSignature, ...quoteWithHash } = deserializedQuote
+  if (
+    !verify(
+      JSON.stringify(quoteWithHash),
+      attesterSignature,
+      deserializedQuote.attesterAddress
+    )
+  ) {
+    throw Error(
+      `attestersSignature ${deserializedQuote.attesterSignature}
+        does not check out with the supplied data`
+    )
+  }
+  const { quoteHash, ...basicQuote } = quoteWithHash
+  if (!validateQuoteSchema(QuoteSchema, basicQuote)) {
+    throw new Error('Quote does not correspond to schema')
+  }
+  if (!verifyQuoteHash(basicQuote, quoteHash)) {
+    throw Error('Invalid Quote Hash')
+  }
+
+  return {
+    ...basicQuote,
+    quoteHash,
+    attesterSignature,
+  }
+}
+
+export function createAttesterSignature(
+  quoteInput: IQuote,
+  attesterIdentity: Identity
+): IQuoteAttesterSigned {
+  const generatedQuoteHash = hashObjectAsStr(quoteInput)
+  const quoteWithHash = { ...quoteInput, quoteHash: generatedQuoteHash }
+  const signature = attesterIdentity.signStr(JSON.stringify(quoteWithHash))
+  if (!verifyQuoteHash(quoteInput, generatedQuoteHash)) {
+    throw Error('Invalid Quote Hash')
+  }
+  return {
+    ...quoteWithHash,
+    attesterSignature: signature,
+  }
+}
+
+export function fromQuoteDataAndIdentity(
+  quoteInput: IQuote,
+  identity: Identity
+): IQuoteAttesterSigned {
+  if (!validateQuoteSchema(QuoteSchema, quoteInput)) {
+    throw new Error('Quote does not correspond to schema')
+  }
+  return createAttesterSignature(quoteInput, identity)
+}
+
+export function createAgreedQuote(
+  claimerIdentity: Identity,
+  attesterSignedQuote: IQuoteAttesterSigned,
+  requestRootHash: string
+): IQuoteAgreement {
+  const { attesterSignature, ...noAttesterSignature } = attesterSignedQuote
+  if (
+    !verify(
+      JSON.stringify({ ...noAttesterSignature }),
+      attesterSignature,
+      attesterSignedQuote.attesterAddress
+    )
+  ) {
+    throw Error(`Quote Signature could not be verified`)
```

Comment:
Change request:
What happens here is different: it _was_ verified, but the signature is invalid. Would be great if the error message would reflect that.

Reply:
I have altered it, what do you think now?